contracts/plugins/assets/RTokenAsset.sol

```diff
@@ -4,57 +4,129 @@ pragma solidity 0.8.9;
+            uint192 q = shiftl_toFix(quantities[i], -int8(IERC20Metadata(erc20s[i]).decimals()));
+
+            // downcast is safe: total attoUoA from any single asset is well under 1e47
+            // D18{UoA} = D18{UoA} + (D18{UoA/tok} * D18{tok} / D18
+            p += uint192((asset.price() * uint256(q)) / FIX_ONE);
+        }
     }
 
-    /// @return max {tok} The maximum trade size
-    function maxTradeSize() external view override returns (uint192 max) {
-        try RTokenPricingLib.price(IRToken(address(erc20))) returns (uint192 p) {
-            // It's correct for the RToken to have a zero price right after a full basket change
-            if (p > 0) {
-                // {tok} = {UoA} / {UoA/tok}
-                // return tradingRange.maxVal.div(p);
-                uint256 max256 = (FIX_ONE_256 * tradingRange.maxVal) / p;
-                if (type(uint192).max < max256) revert UIntOutOfBounds();
-                max = uint192(max256);
-            }
+    /// @return p {UoA/tok} The current price(), or if it's reverting, a fallback price
+    function priceWithFailover() public view virtual returns (uint192 p) {
+        // solhint-disable no-empty-blocks
+        try this.price() returns (uint192 price_) {
+            p = price_;
         } catch {}
-        if (max == 0 || max > tradingRange.maxAmt) max = tradingRange.maxAmt;
-        if (max < tradingRange.minAmt) max = tradingRange.minAmt;
+        // solhint-enable no-empty-blocks
+
+        if (p == 0) {
+            IMain main = IRToken(address(erc20)).main();
+            IAssetRegistry assetRegistry = main.assetRegistry();
+            uint256 totalSupply = IRToken(address(erc20)).totalSupply();
+            uint256 basketsNeeded = IRToken(address(erc20)).basketsNeeded();
+
+            uint192 amtBUs; // D18{BU}
+            if (totalSupply > 0) {
+                // Can allow the totalSupply to be zero if we are getting the fallback price
+                amtBUs = uint192((basketsNeeded * FIX_ONE_256) / totalSupply);
+            }
+
+            (address[] memory erc20s, uint256[] memory quantities) = main.basketHandler().quote(
+                amtBUs,
+                CEIL
+            );
+
+            uint256 erc20length = erc20s.length;
+
+            // Bound each withdrawal by the prorata share, in case we're currently under-capitalized
+            for (uint256 i = 0; i < erc20length; ++i) {
+                IAsset asset = assetRegistry.toAsset(IERC20(erc20s[i]));
+
+                // D18{tok} = D18 * {qTok} / {qTok/tok}
+                uint192 q = shiftl_toFix(
+                    quantities[i],
+                    -int8(IERC20Metadata(erc20s[i]).decimals())
+                );
+
+                // downcast is safe: total attoUoA from any single asset is well under 1e47
+                // D18{UoA} = D18{UoA} + (D18{UoA/tok} * D18{tok} / D18
+                p += uint192((asset.priceWithFailover() * uint256(q)) / FIX_ONE);
+            }
+        }
```

Comment:
This seems fine, but there's heavy repetition of code. This could be two entry points to a single parameterized function, with just a bit like `uint192 assetPrice = failover?asset.priceWithFailover():asset.price()`

Reply:
Notably, you don't then need the outermost try-catch in priceWithFailover().

Reply:
I'm aware, I tried it both ways. There are 3 differences to hit:
1. The CEIL vs FLOOR difference in the `quote` call
2. The prorata calculation
3. `asset.priceWithFailover()` vs `asset.price()`

which was just enough to get me to revert it to the fully written out thing for both functions.

Reply:
oh, and also a 4th: how `totalSupply == 0` case is treated.